pdns/recursordist/rec-system-resolve.cc

```diff
@@ -0,0 +1,292 @@
+    if (iter->second.d_ttd < now) {
+      return iter->second.d_address;
+    }
+    // If it's stale, re-resolve below
+  }
+  // We keep the lock while resolving, even though this might take a while...
+  auto address = resolve(name);
+
+  time_t ttd = now + d_ttl;
+  auto iter = data->d_map.emplace(name, AddressData{address, ttd}).first;
+  return iter->second.d_address;
+}
+
+ComboAddress pdns::RecResolve::lookup(const std::string& name)
+{
+  auto data = d_data.lock();
+  if (auto iter = data->d_map.find(name); iter != data->d_map.end()) {
+    // always return it, even if it's stale
+    return iter->second.d_address;
+  }
+  throw PDNSException("system resolve of unregistered name: " + name);
+}
+
+void pdns::RecResolve::wipe(const string& name)
+{
+  auto data = d_data.lock();
+  if (name.empty()) {
+    data->d_map.clear();
+  }
+  else {
+    data->d_map.erase(name);
+  }
+}
+
+bool pdns::RecResolve::refresh(time_t now)
+{
+  // The refresh task should not take the lock for a long time, so we're working on a copy
+  ResolveData copy;
+  {
+    auto data = d_data.lock();
+    copy = *data;
+  }
+  std::map<std::string, AddressData> newData;
+
+  auto log = g_slog->withName("system-resolver");
+
+  bool updated = false;
+  for (const auto& entry : copy.d_map) {
+    if (entry.second.d_ttd <= now) {
+      auto newAddress = resolve(entry.first);
+      time_t ttd = now;
+      if (newAddress != ComboAddress()) {
+        // positive resolve, good for ttl
+        ttd += d_ttl;
+      }
+      else {
+        log->error(Logr::Error, "Name did not resolve", "name", Logging::Loggable(entry.first));
+      }
+      if (newAddress != entry.second.d_address) {
+        log->info(Logr::Debug, "Name resolved to new address",
+                  "name", Logging::Loggable(entry.first),
+                  "address", Logging::Loggable(newAddress.toString()));
+        // An address changed
+        updated = true;
+      }
+      newData.emplace(entry.first, AddressData{newAddress, ttd});
+    }
+  }
+
+  if (!newData.empty()) {
+    auto data = d_data.lock();
+    for (const auto& entry : newData) {
+      data->d_map.insert_or_assign(entry.first, entry.second);
+    }
+  }
+  if (updated) {
+    log->info(Logr::Info, "Changes in names detected");
+  }
+  return updated;
+}
+
+bool pdns::RecResolve::changeDetected()
+{
+  bool change = d_refresher.changes.exchange(false);
+  return change;
+}
+
+pdns::RecResolve::Refresher::Refresher(time_t interval, const std::function<void()>& callback, pdns::RecResolve& res) :
+  d_resolver(res), d_callback(callback), d_interval(std::max(static_cast<time_t>(1), interval))
+{
+  start();
+}
+
+pdns::RecResolve::Refresher::~Refresher()
+{
+  finish();
+}
+
+void pdns::RecResolve::Refresher::refreshLoop()
+{
+  setThreadName("rec/sysres");
+  time_t lastSelfCheck = 0;
+
+  while (!stop) {
+    const time_t startTime = time(nullptr);
+    time_t wakeTime = startTime;
+    while (wakeTime - startTime < d_interval) {
```

Comment:
This looks strange to me. Wouldn't `wakeTime - startTime` be zero?

Reply:
yes it would initially be zero, the expression `wakeTime - startTime`  is equal to the time slept. I'll add a comment